### PR TITLE
MERGE? np backend/frontend

### PR DESCRIPTION
**Pull request by edouardoyallon**

### CODE REVIEW

File: kymatio/scattering3d/backend/numpy_backend.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+from collections import namedtuple
+from scipy.fftpack import fftn, ifftn
+
+
+def _iscomplex(x):
+    return x.dtype == np.complex64 or x.dtype == np.complex128
+
+
+def complex_modulus(input_array):
+    """Computes complex modulus.
```
Comment (janden) [inline kymatio/scattering3d/backend/numpy_backend.py]: “tensor” should be array and the output is no longer “complex”.

File: kymatio/scattering3d/backend/numpy_backend.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+            List of P positive floats containing the p values used to
+            compute the integrals of the input_array to the power p (l_p
+            norms).
+        Returns
+        -------
+        integrals : torch tensor
+            Tensor of size (B, P) containing the integrals of the input_array
+            to the powers p (l_p norms).
+    """
+    integrals = np.zeros((input_array.shape[0], len(integral_powers)),dtype=np.complex64)
```
Comment (janden) [inline kymatio/scattering3d/backend/numpy_backend.py]: Space before `dtype`.

File: kymatio/scattering3d/core/scattering3d.py [hunk 3]
```diff
@@ -41,7 +41,7 @@ def scattering3d(x, filters, rotation_covariant, L, J, max_order, backend, avera
                 U_1_m = modulus(U_1_c)
 
             U_1_c = fft(U_1_m)
-            S_1_l = averaging(U_1_c, j_1)[..., 0]
+            S_1_l = averaging(U_1_c, j_1)
```
Comment (janden) [inline kymatio/scattering3d/core/scattering3d.py]: This will break the Torch code.

File: kymatio/scattering3d/frontend/numpy_frontend.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        Returns
+        -------
+        output: tuple | numpy array
+            if max_order is 1 it returns a numpy array with the
+            first order scattering coefficients
+            if max_order is 2 it returns a numpy array with the
+            first and second order scattering coefficients,
+            concatenated along the feature axis
+        """
+        if (input_array.shape[-1] != self.O or input_array.shape[-2] != self.N
```
Comment (janden) [inline kymatio/scattering3d/frontend/numpy_frontend.py]: Need to make sure it's (at least) 3D (like in Torch frontend).

File: kymatio/scattering3d/frontend/numpy_frontend.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            or input_array.shape[-3] != self.M):
+            raise RuntimeError(
+                'Tensor must be of spatial size (%i, %i, %i).' % (
+                    self.M, self.N, self.O))
+
+        batch_shape = input_array.shape[:-3]
+        signal_shape = input_array.shape[-3:]
+
+        input_array = input_array.reshape((-1,)+signal_shape)
```
Comment (janden) [inline kymatio/scattering3d/frontend/numpy_frontend.py]: Spaces around `+`.

File: kymatio/scattering3d/tests/test_numpy_scattering3d.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+backends.append(backend)
+
+
+def relative_difference(a, b):
+    return np.sum(np.abs(a - b)) / max(np.sum(np.abs(a)), np.sum(np.abs(b)))
+
+
+@pytest.mark.parametrize("backend", backends)
+def test_against_standard_computations(backend):
+
```
Comment (janden) [inline kymatio/scattering3d/tests/test_numpy_scattering3d.py]: No blank line.

File: kymatio/scattering3d/tests/test_numpy_scattering3d.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                                      integral_powers=integral_powers,
+                                      max_order=2, backend=backend,
+                                      frontend='numpy')
+
+    order_0 = backend.compute_integrals(x, integral_powers)
+    scattering.max_order = 2
+    scattering.method = 'integral'
+    scattering.integral_powers = integral_powers
+    orders_1_and_2 = scattering(x)
+    #Jprint(orders_1_and_2)
```
Comment (janden) [inline kymatio/scattering3d/tests/test_numpy_scattering3d.py]: Remove.

File: kymatio/scattering3d/tests/test_numpy_scattering3d.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    order_0_diff_cpu = relative_difference(order_0_ref, order_0)
+    print(orders_1_and_2_ref.shape)
+    print(orders_1_and_2.shape)
+    orders_1_and_2_diff_cpu = relative_difference(
+        orders_1_and_2_ref, orders_1_and_2)
+
+    assert order_0_diff_cpu < 1e-6, "CPU : order 0 do not match, diff={}".format(order_0_diff_cpu)
+    assert orders_1_and_2_diff_cpu < 1e-6, "CPU : orders 1 and 2 do not match, diff={}".format(orders_1_and_2_diff_cpu)
+
```
Comment (janden) [inline kymatio/scattering3d/tests/test_numpy_scattering3d.py]: Remove blank line. Add test for batch shape agnosticism test (in current 0.2.X branch for torch).

File: kymatio/scattering3d/backend/numpy_backend.py (hunk 1 above)
Comment (MuawizChaudhary) [reply kymatio/scattering3d/backend/numpy_backend.py]: Docathon

File: kymatio/scattering3d/core/scattering3d.py (hunk 3 above)
Comment (MuawizChaudhary) [reply kymatio/scattering3d/core/scattering3d.py]: I have a separate PR that fixes this.

File: kymatio/scattering3d/frontend/numpy_frontend.py (hunk 4 above)
Comment (MuawizChaudhary) [reply kymatio/scattering3d/frontend/numpy_frontend.py]: Does this check not do that?

File: kymatio/scattering3d/tests/test_numpy_scattering3d.py (hunk 8 above)
Comment (MuawizChaudhary) [reply kymatio/scattering3d/tests/test_numpy_scattering3d.py]: I may need someone to look over the batch agnosticism test before I can proceed.

File: kymatio/scattering3d/tests/test_numpy_scattering3d.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    orders_1_and_2_ref = scattering_ref[:, start:end]
+
+    order_0_diff_cpu = relative_difference(order_0_ref, order_0)
+    print(orders_1_and_2_ref.shape)
+    print(orders_1_and_2.shape)
+    orders_1_and_2_diff_cpu = relative_difference(
+        orders_1_and_2_ref, orders_1_and_2)
+
+    assert order_0_diff_cpu < 1e-6, "CPU : order 0 do not match, diff={}".format(order_0_diff_cpu)
+    assert orders_1_and_2_diff_cpu < 1e-6, "CPU : orders 1 and 2 do not match, diff={}".format(orders_1_and_2_diff_cpu)
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/tests/test_numpy_scattering3d.py]: Add batch shape agnosticism test.

File: kymatio/scattering3d/frontend/numpy_frontend.py (hunk 4 above)
Comment (janden) [reply kymatio/scattering3d/frontend/numpy_frontend.py]: No it errors if `shape` is too short.

File: kymatio/scattering3d/frontend/numpy_frontend.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+            first and second order scattering coefficients,
+            concatenated along the feature axis
+        """
+        if not type(input_array) is np.ndarray:
+            raise TypeError('The input should be a NumPy array.')
+        
+        if input_array.ndim < 3:
+            raise RuntimeError('Input tensor must have at least three '
+                               'dimensions.')
+        print(input_array.ndim)
```
Comment (MuawizChaudhary) [inline kymatio/scattering3d/frontend/numpy_frontend.py]: Remove.

**Pull request merged**